### PR TITLE
Optimizer: Clear state for JUMPDESTs.

### LINKED ISSUE #1416: Optimizer causes code that overwrites contract fields, with an if(false) nearby
> Which version of Solidity you are using

solc is 0.4.5+commit.b318366e.Linux.g++

> What was the source code (if applicable)

```
pragma solidity ^0.4.5;

contract Test {
    uint256 public totalSupply;
    function Test() {
        totalSupply = 1000;
    }
    function f() {
        if (false) // comment the if out to fix
            var other_name = totalSupply;
        totalSupply -= 456;
    }
}
```

> Which platform are you running on

populus (TestRPC), online solidity compiler as well

> How to reproduce the issue

compile with optimization, deploy `Test`, call `f`, `totalSupply` is 2^256 - 456 afterwards

populus:
```
def test_test(chain):
    contract_factory = chain.get_contract_factory("Test")
    deploy_txn_hash = contract_factory.deploy()
    deploy_address = chain.wait.for_contract_address(deploy_txn_hash)
    asset = contract_factory(address=deploy_address)

    assert asset.call().totalSupply() == 1000
    txn_hash = asset.transact().f()
    chain.wait.for_receipt(txn_hash)
    assert asset.call().totalSupply() == 544
```

> What was the result of the issue

with the `if(false)` condition in the contract, the `totalSupply` variable gets overwritten with 0 somewhere.

> What the expected behaviour is

`totalSupply == 544`

### PR DESCRIPTION
**Pull request by chriseth**
Fixes #1416

### CODE REVIEW

File: libevmasm/Assembly.cpp [hunk 1]
```diff
@@ -360,46 +360,35 @@ map<u256, u256> Assembly::optimiseInternal(bool _enable, bool _isCreation, size_
...
-						break;
-
 				KnownState emptyState;
 				CommonSubexpressionEliminator eliminator(emptyState);
-				auto blockIter = iter;
-				auto const blockEnd = end;
-				while (blockIter < blockEnd)
+				auto orig = iter;
+				iter = eliminator.feedItems(iter, m_items.end());
+				bool shouldReplace = false;
```
Comment (chriseth) [inline libevmasm/Assembly.cpp:367]: Mostly only indentation change from here.

File: test/libsolidity/SolidityOptimizer.cpp [hunk 2]
```diff
@@ -1246,6 +1246,26 @@ BOOST_AUTO_TEST_CASE(dead_code_elimination_across_assemblies)
...
+		contract Test {
+			uint256 public totalSupply = 100;
+			function f() returns (uint r) {
+				if (false)
+					r = totalSupply;
+				totalSupply -= 10;
+			}
+			function test() returns (uint) {
+				f();
+				return this.totalSupply();
```
Comment (axic) [inline test/libsolidity/SolidityOptimizer.cpp:1261]: Does this has to be `this.` to trigger the bug?

Comment (chriseth) [reply test/libsolidity/SolidityOptimizer.cpp:1261]: No, this is just the "external testing function". The bug is triggered by unreachable code in the `if (false)`

**Review by pirapira — APPROVED** [review]
I asked @chriseth some questions till I got the change.

**Pull request merged**